---
title: "CosMx Pancreas Analysis"
author: "Ash Fletcher"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE,
                      fig.width = 10, fig.height = 8, fig.align = "center")
```

# Introduction

This document contains the analysis of CosMx spatial transcriptomics data from pancreatic tissue. 
The analysis includes data processing, visualization, and differential expression analysis.

```{r install_packages, include=FALSE}
# Install required packages if not already installed
if (!require("Seurat")) install.packages("Seurat")
if (!require("tidyverse")) install.packages("tidyverse")
if (!require("ggplot2")) install.packages("ggplot2")
if (!require("future")) install.packages("future")
if (!require("EnhancedVolcano")) install.packages("EnhancedVolcano")
if (!require("glmGamPoi")) {
  if (!require("BiocManager")) install.packages("BiocManager")
  BiocManager::install("glmGamPoi")
}
```

# Libraries

```{r libraries}
library(Seurat)
library(tidyverse)
library(glmGamPoi)
library(ggplot2)
library(future)
library(EnhancedVolcano)

options(future.globals.maxSize = 8000 * 1024^2)
```

# Plot colors

```{r helper_functions}
# Define color palette using glasbey and polychrome colors
colors_20 <- DiscretePalette(20, palette = "glasbey", shuffle = FALSE)
colors_10 <- DiscretePalette(10, palette = "polychrome", shuffle = FALSE)
```

# Read in data

```{r read data}
so_panc <- readRDS("so_panc_cosmxWTA_sct.rds")
Idents(so_panc) <- "types"
```


# Data Visualization

## Spatial Visualization {.tabset .tabset-fade}

### Single FOV

```{r single_fov, fig.width=10, fig.height=8}
# Zoomed to single FOV: CellOverlay_F053.jpg
# Visualize cell types in zoom 1
ImageDimPlot(object = so_panc, fov = "zoom1", group.by = "types", axes = TRUE, cols = "glasbey")
```

### Multiple FOVs

```{r multiple_fovs, fig.width=10, fig.height=8}
# Zoomed to 4 FOVs: CellOverlay_F051.jpg - CellOverlay_F054.jpg
# Visualize cell types in zoom 4
ImageDimPlot(object = so_panc, fov = "zoom4", group.by = "types", axes = TRUE, cols = "glasbey")
```

## Gene Expression {.tabset .tabset-fade}

```{r}
molsAll <- c("GCG", "EPCAM", "PRSS1", "C1QA",
          "ACTA2", "CD3E", "INS", "CD163", "CD86")
molsDuct <- c("KRT19", "SOX9", "CFTR", "MUC1")
molsBeta <- c("INS", "IAPP", "DLK1", "PDX1")
molsMacs <- c("CD68", "CD163", "CD86", "CSF1R", "MARCO")
molsPSC <- c("FAP", "ACTA2", "COL1A1", "PDGFRA")
```

### All Genes

```{r molecule_visualization, fig.width=10, fig.height=8}
# Plot molecules in single zoom segment
ImageDimPlot(so_panc, fov = "zoom1", cols = "polychrome", alpha = 0.25, molecules = molsAll, 
             mols.size = 0.35, nmols = 20000, border.color = "black", coord.fixed = FALSE)
```

### Ductal Markers

```{r ductal_markers, fig.width=10, fig.height=8}
ImageDimPlot(so_panc, fov = "zoom1", cols = "polychrome", alpha = 0.25, molecules = molsDuct, 
             mols.size = 0.35, nmols = 20000, border.color = "black", coord.fixed = FALSE)
```

### T Cell Markers

```{r tcell_markers, fig.width=10, fig.height=8}
ImageDimPlot(so_panc, fov = "zoom1", cols = "polychrome", alpha = 0.25, molecules = molsBeta, 
             mols.size = 0.35, nmols = 20000, border.color = "black", coord.fixed = FALSE)
```

### Macrophage Markers

```{r macrophage_markers, fig.width=10, fig.height=8}
ImageDimPlot(so_panc, fov = "zoom1", cols = "polychrome", alpha = 0.25, molecules = molsMacs, 
             mols.size = 0.35, nmols = 20000, border.color = "black", coord.fixed = FALSE)
```

### PSC Markers

```{r psc_markers, fig.width=10, fig.height=8}
ImageDimPlot(so_panc, fov = "zoom1", cols = "polychrome", alpha = 0.25, molecules = molsPSC, 
             mols.size = 0.35, nmols = 20000, border.color = "black", coord.fixed = FALSE)
```

## Cell Type Specific Genes in Specific Cells

```{r molecule_visualization specific cells, fig.width=10, fig.height=8}
ImageDimPlot(so_panc, fov = "zoom1", cells = WhichCells(so_panc, 
            idents = c("Acinar.1", "Acinar.2", "Ductal","Macrophage")), 
             cols = colors_10, size = 0.6,
             molecules = c("MUC1", "SPINK1", "CD14"),
             alpha = 0.35, mols.size = 0.7, mols.cols = c("blue", "green", "yellow"))
```

# Single-Cell Analysis

## UMAP Visualization {.tabset}

### Visualize by cluster
```{r single_cell_analysis clusters, fig.width=10, fig.height=8}
# Visualize cluster UMAP
DimPlot(so_panc, group.by = "seurat_clusters") + scale_color_manual(values = colors_20)
```

### Visualize by cell type
```{r single_cell_analysis cells, fig.width=10, fig.height=8}
# Visualize cell type UMAP
DimPlot(so_panc) + scale_color_manual(values = colors_20)
```

## Gene Expression Plots {.tabset}

### Violin plot
```{r gene_expression violin, fig.width=10, fig.height=8}
# Gene expression visualization
VlnPlot(so_panc, features = c("VIM", "PRSS1"), assay = "SCT", 
        layer = "counts", pt.size = 0.1) + NoLegend()
```

### Feature plot
```{r gene_expression feature, fig.width=10, fig.height=8}
FeaturePlot(so_panc, features = c("KRT17", "VIM", "PRSS1", "CD163"), 
            max.cutoff = "q95")
```

# Differential Expression Analysis

## Find Markers

```{r differential_expression find markers, fig.width=10, fig.height=8}
# Find markers between Acinar.1 and Acinar.2 ## This took me ~8mins on my laptop
acini_markers <- FindMarkers(so_panc, ident.1 = "Acinar.1",
                            ident.2 = "Acinar.2", verbose = TRUE, 
                            assay = "SCT")
```

## Volcano Plot

```{r differential_expression volcano, fig.width=10, fig.height=8}
# Volcano plot parameters
fc <- 1
pval <- 0.05

degs <- acini_markers %>%
  filter(abs(avg_log2FC) > fc) %>% 
  filter(p_val_adj <= pval) 

# Create volcano plot
EnhancedVolcano(acini_markers,
                lab = rownames(acini_markers),
                selectLab = rownames(degs),
                subtitle = bquote(italic("Acinar.2 <---> Acinar.1")),
                x = 'avg_log2FC',
                y = 'p_val_adj',
                legendLabels = c("NS", paste0("Log2FC >", fc),
                                paste0("Adj. p-val <", pval), 
                                paste0("Adj. p-val <", pval, " and Log2FC >", fc)),
                subtitleLabSize = 15,
                captionLabSize = 15,
                pCutoff = pval,
                FCcutoff = fc,
                pointSize = 2.5,
                labSize = 4,
                axisLabSize = 14,
                legendLabSize = 14, 
                colAlpha = 0.35)
```

# Interactive Exploration

## See if/where your favorite genes show up in the pancreas

```{r gene explore, fig.width=10, fig.height=8}
my_genes <- c("SOX9")

ImageDimPlot(so_panc, fov = "zoom4", cols = "polychrome", alpha = 0.3, molecules = my_genes, 
             mols.size = 0.35, nmols = 20000, border.color = "black", coord.fixed = FALSE)
```
